secret/secret.go

```diff
@@ -64,48 +68,147 @@ func UpdateSecrets(s secretInterface, secrets *v1.Secret, create bool) {
+			// the V of the current best.
+			//
+			// - A longer name indicates a longer V, that
+			//   is larger (*). This relies on the `no
+			//   leading 0's` part.
+			//
+			// - For V's of the same length the lexico-
+			//   graphical order is identical to the
+			//   numeric order. This also relies on the
+			//   `no leading 0's` part.
+			//
+			// (*) The special case where the name is just
+			// `foo` is handled by this as well, as a
+			// zero-length V which is less than anything
+			// else and thus only used when nothing else
+			// is present, in line with `foo` being the
+			// last resort fallback.
+
+			best = s
+		}
 	}
+
+	if best.Name != "" {
+		return &best, nil
+	}
+	return nil, errors.NewNotFound(v1.Resource("secret"), SECRET_NAME)
 }
 
 func GetOrCreateSecrets(s secretInterface) (create bool, secrets, updates *v1.Secret) {
 	secretUpdateName := SECRET_UPDATE_NAME
+	secretName := SECRET_NAME
 	releaseRevision := getEnv("RELEASE_REVISION")
 	if releaseRevision != "" {
 		secretUpdateName += "-" + releaseRevision
+		secretName += "-" + releaseRevision
 	}
 
+	log.Printf("Checking for chart-provided `%s`\n", secretUpdateName)
+
 	// secret updates *must* exist
 	updates, err := s.Get(secretUpdateName, metav1.GetOptions{})
 	if err != nil {
 		logFatal(err)
 		return false, nil, nil
 	}
 
+	log.Println("Now ready to generate secrets")
+
 	// check for existing secret, initialize a new Secret if not found
-	secrets, err = s.Get(SECRET_NAME, metav1.GetOptions{})
+	secrets, err = FindSecret(s)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			log.Println("`secret` not found, creating")
-			create = true
+			log.Printf("No previous secret found, will create `%s`\n",
+				secretName)
 
+			create = true
 			secrets = &v1.Secret{
 				ObjectMeta: metav1.ObjectMeta{
-					Name: SECRET_NAME,
+					Name: secretName,
 				},
 				Data: map[string][]byte{},
 			}
 		} else {
 			logFatal(err)
 			return false, nil, nil
 		}
+	} else {
+		if secrets.Name != secretName {
+			// The structure was imported and the
+			// destination to write to is different than
+			// what we got (per the names). Force creation.
+
+			log.Printf("Imported previous secret `%s`, will create `%s`\n",
+				secrets.Name, secretName)
+
+			create = true
+
+			// Note that the imported structure has its
+			// field `ResourceVersion` set. That makes it
+			// unsuitable for use with `Create`.  `Create`
+			// actually rejects it and aborts the process.
+
+			// The field is noted as read-only, we cannot
+			// unset it.  We deal with the situation by
+			// creating a fresh structure, i.e. without
+			// `ResourceVersion` set, and give it a copy
+			// of the imported data (and the destination
+			// name, of course).
+			newsecrets := &v1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: secretName,
+				},
+				Data: secrets.Data,
+			}
+			secrets = newsecrets
+		} else {
+			log.Printf("Imported secret `%s`, will update\n",
```

Comment:
I would put some text here that this situation can only happen when `$RELEASE_REVISION` has not been set, which should only ever be true when running the tests, but not when used in real usage.

Reply:
Oh, you mean the unit tests ? Would it not be better then to have the unit tests mock up a revision ?
See my note above too about RV.

Reply:
Yes. agreed.

Reply:
Will do.